src/Compiler.php

```diff
@@ -24,55 +24,84 @@
 
 namespace Datto\Cinnabari;
 
-use Datto\Cinnabari\Exception\CompilerException;
-use Datto\Cinnabari\Compiler\GetCompiler;
 use Datto\Cinnabari\Compiler\DeleteCompiler;
+use Datto\Cinnabari\Compiler\GetCompiler;
+use Datto\Cinnabari\Compiler\SetCompiler;
+use Datto\Cinnabari\Compiler\InsertCompiler;
+use Datto\Cinnabari\Exception\CompilerException;
 
 /**
  * Class Compiler
  * @package Datto\Cinnabari
  */
 class Compiler
 {
+    const TYPE_UNKNOWN = -1;
     const TYPE_GET = 0;
     const TYPE_DELETE = 1;
+    const TYPE_SET = 2;
+    const TYPE_INSERT = 3;
 
     private $getCompiler;
     private $deleteCompiler;
+    private $setCompiler;
+    private $translator;
 
-    public function __construct()
+    public function __construct($schema)
     {
         $this->getCompiler = new GetCompiler();
         $this->deleteCompiler = new DeleteCompiler();
+        $this->setCompiler = new SetCompiler();
+        $this->insertCompiler = new InsertCompiler();
+        $this->translator = new Translator($schema);
     }
     
-    public function compile($translatedRequest, $arguments)
+    public function compile($request, $arguments)
     {
-        $queryType = self::getQueryType($translatedRequest);
+        $queryType = self::getQueryType($request);
 
         switch ($queryType) {
             case self::TYPE_GET:
+                $translatedRequest = $this->translator->translateIgnoringObjects($request);
                 return $this->getCompiler->compile($translatedRequest, $arguments);
 
             case self::TYPE_DELETE:
+                $translatedRequest = $this->translator->translateIgnoringObjects($request);
                 return $this->deleteCompiler->compile($translatedRequest, $arguments);
+
+            case self::TYPE_SET:
+                $translatedRequest = $this->translator->translateIncludingObjects($request);
+                return $this->setCompiler->compile($translatedRequest, $arguments);
+
+            case self::TYPE_INSERT:
+                $translatedRequest = $this->translator->translateIncludingObjects($request);
+                return $this->insertCompiler->compile($translatedRequest, $arguments);
     
             default:
-                throw CompilerException::unknownRequestType($translatedRequest);
+                throw CompilerException::unknownRequestType($request);
         }
     }
 
-    public static function getQueryType($translatedRequest)
+    public static function getQueryType($request)
     {
-        $lastRequest =  end($translatedRequest);
-        list($lastTokenType, $lastToken) = each($lastRequest);
+        list($tokenType, $functionName, ) = reset($request);
```

Comment:
What if the user provides a request that doesn't start with a function token? What if the token array contains fewer elements?